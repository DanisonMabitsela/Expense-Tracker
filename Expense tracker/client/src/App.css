/* CSS for Sidebar */
.sidebar {
  width: 250px;
  background-color: #f1f1f1;
  position: fixed;
  left: 0;
  top: 0;
  height: 100%;
  overflow-y: auto;
}

.sidebar ul {
  list-style-type: none;
  padding: 0;
}

.sidebar li {
  margin: 10px 0;
}

.sidebar a {
  display: block;
  padding: 15px;
  text-decoration: none;
  color: #000;
}

.sidebar a:hover {
  background-color: #ddd;
}

/* CSS for App */
.app {
  display: flex;
}

.app .content {
  flex: 1;
  padding: 20px;
  margin-left: 250px;
}

/* Custom styles for your theme */
body {
  background-color: #e2e2ee;
}

.sidebar {
  background-color: #78e386;
}

.sidebar a {
  color: #555;
}

.sidebar a:hover {
  background-color: #ddd;
  color: #000;
}

/* Responsive Styles */
@media only screen and (max-width: 768px) {
  .sidebar {
    width: 100%;
    height: auto;
    position: relative;
    left: initial;
    top: initial;
    margin-bottom: 20px;
  }

  .app .content {
    margin-left: 0;
  }
}
