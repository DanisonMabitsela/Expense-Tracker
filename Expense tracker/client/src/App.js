import React, { useState, useEffect } from "react";
import { BrowserRouter as Router, Route, Routes, Link } from "react-router-dom";
import Register from "./components/Register";
import Login from "./components/Login";
import CreateExpense from "./components/CreateExpense";
import CreateIncome from "./components/CreateIncome";
import ExpenseList from "./components/ExpenseList";
import IncomeList from "./components/IncomeList";
import BalanceSummary from "./components/BalanceSummary";
import AdminComponent from "./components/AdminComponent"; // Import AdminComponent
import "./App.css";

const Sidebar = ({ onLogout, isAdmin }) => {
  return (
    <div className="sidebar">
      <ul className="sidebar-list">
        <li>
          <Link to="/" className="sidebar-link">
            Home
          </Link>
        </li>
        <li>
          <Link to="/create-expense" className="sidebar-link">
            Create Expense
          </Link>
        </li>
        <li>
          <Link to="/create-income" className="sidebar-link">
            Create Income
          </Link>
        </li>
        <li>
          <Link to="/expenses" className="sidebar-link">
            Expenses
          </Link>
        </li>
        <li>
          <Link to="/incomes" className="sidebar-link">
            Incomes
          </Link>
        </li>
        {isAdmin && ( // Render link to AdminComponent if user is admin
          <li>
            <Link to="/user" className="sidebar-link">
              user
            </Link>
          </li>
        )}
      </ul>
      <li>
        <button onClick={onLogout} className="logout-button">
          Logout
        </button>
      </li>
    </div>
  );
};

const App = () => {
  const [token, setToken] = useState(sessionStorage.getItem("token"));
  const [isAdmin, setIsAdmin] = useState(false);

  // Function to check if user is admin
  const checkIsAdmin = async () => {
    if (token) {
      try {
        const response = await fetch("http://localhost:5000/api/user", {
          headers: {
            Authorization: token,
          },
        });
        const data = await response.json();
        setIsAdmin(data.isAdmin);
      } catch (error) {
        console.error("Error fetching user data:", error);
      }
    }
  };

  useEffect(() => {
    checkIsAdmin(); // Call checkIsAdmin on mount and when token changes
  }, [token]);

  const handleLogout = () => {
    sessionStorage.clear();
    setToken(null);
    setIsAdmin(false); // Reset isAdmin state to false
  };

  return (
    <Router>
      <div className="app">
        {token && (
          <Sidebar
            onLogout={handleLogout}
            isAdmin={isAdmin}
            checkIsAdmin={checkIsAdmin}
          />
        )}
        <Routes>
          <Route
            path="/"
            element={
              token ? (
                <BalanceSummary token={token} />
              ) : (
                <Login setToken={setToken} setIsAdmin={setIsAdmin} />
              )
            }
          />
          <Route path="/register" element={<Register />} />
          <Route
            path="/login"
            element={<Login setToken={setToken} setIsAdmin={setIsAdmin} />}
          />
          <Route
            path="/create-expense"
            element={token ? <CreateExpense token={token} /> : <Login />}
          />
          <Route
            path="/create-income"
            element={token ? <CreateIncome token={token} /> : <Login />}
          />
          <Route
            path="/expenses"
            element={
              token ? (
                <ExpenseList token={token} isAdmin={isAdmin} />
              ) : (
                <Login />
              )
            }
          />
          <Route
            path="/incomes"
            element={
              token ? <IncomeList token={token} isAdmin={isAdmin} /> : <Login />
            }
          />
          {isAdmin && ( // Render route to AdminComponent if user is admin
            <Route
              path="/user"
              element={
                <AdminComponent
                  isAdmin={isAdmin}
                  checkIsAdmin={checkIsAdmin}
                  token={token} // Pass token as a prop to AdminComponent
                />
              }
            />
          )}
        </Routes>
      </div>
    </Router>
  );
};

export default App;
