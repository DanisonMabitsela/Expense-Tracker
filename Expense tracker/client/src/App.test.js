import React from "react";
import { render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import App from "./App";

it("renders correctly", () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("App", () => {
  it("logs out the user when the logout button is clicked", () => {
    // Set up the test
    sessionStorage.setItem("token", "test-token");
    const { getByText } = render(<App />);
    const logoutButton = getByText(/logout/i);

    // Verify that the token is set in local storage
    expect(sessionStorage.getItem("token")).toBe("test-token");

    // Simulate clicking the logout button
    fireEvent.click(logoutButton);

    // Verify that the token is removed from local storage
    expect(sessionStorage.getItem("token")).toBeNull();
  });

  it("fetches user data when the token is set", async () => {
    // Set up the test
    sessionStorage.setItem("token", "test-token");
    const mockFetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve({ isAdmin: true }),
      })
    );
    global.fetch = mockFetch;
    render(<App />);

    // Verify that the fetch function was called with the correct arguments
    expect(mockFetch).toHaveBeenCalledWith("http://localhost:5000/api/user", {
      headers: {
        Authorization: "test-token",
      },
    });

    // Clean up
    global.fetch.mockRestore();
  });
});
