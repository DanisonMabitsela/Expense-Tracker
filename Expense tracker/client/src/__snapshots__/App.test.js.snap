// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="app"
>
  <div
    className="login-container"
  >
    <h2>
      Login
    </h2>
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
      <input
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        type="submit"
      >
        Login
      </button>
    </form>
    <button
      onClick={[Function]}
    >
      Register
    </button>
  </div>
</div>
`;
